### ISSUE TITLE
Missing `query()` method on Connection - intentional?

### ISSUE THREAD

**Issue by tspence** [jgallagher/rusqlite#489]
On the connection object, there are methods such as `query_row` that specifies that only one row is returned.  However, to execute a query that returns multiple rows, we must first prepare a statement and then call query on it.

Since the implementation of Connection.query_row() internally prepares a statement then returns the first row of the result, is there any reason we shouldn't have a `query` method that returns all rows on the connection object?

**Comment by thomcc** [comment]
This is probably worth adding. In https://github.com/mozilla/application-services, we have a wrapper that helps with this https://github.com/mozilla/application-services/blob/432f399c30812b397c189ad017f6c3ad1a3ff077/components/support/sql/src/conn_ext.rs#L148-L163, as it's often convenient (Admittedly, even if this were in rusqlite directly we'd probably still use the wrapper since it allows for easy caching and logs query plan info depending on `#[cfg(feature)]`, etc).

**Comment by tspence** [comment]
Thank you for the quick response!

Just from an outsiders' perspective, I would think that "method returning all rows" would be the normal default case worth implementing, and "method returning just one single row" would be the special case you could decline to support to avoid excessive methods.

In general, capturing only one row would be very easy to implement by retrieving multiple rows then closing the row object after retrieving one, or by calling a multiple row function using `LIMIT 1`.

**Comment by gwenn** [comment]
I am not sure that #478 will be merged (major breaking change).
But, if merged, we can give access to a [FallibleStreamingIterator](https://docs.rs/fallible-streaming-iterator/0.1.9/fallible_streaming_iterator/index.html) from a connection object.
Or wait until there is a streaming iterator/generator in the standard library.

**Comment by thomcc** [comment]
> In general, capturing only one row would be very easy to implement by retrieving multiple rows then closing the row object after retrieving one, or by calling a multiple row function using LIMIT 1.

Right, but you still need all the cases I described just to implement the method returning all rows (or to make a decision about which ones are worth supporting. At a minimum this means deciding named vs positional parameters, and if your mapper function returns T or Result<T, E>.

(A FallibleStreamingIterator solves case 2, I think, but not case 1)

**Comment by tspence** [comment]
Thank you for the detailed response.  I did a quick fork to see what the code looks like, and I think I see why returning a multi-row object is not as trivial as I thought from the outside.

Regarding "positional vs named" and "query_map" vs "query_and_then" - I do see that you're supporting a lot of different convenience cases in the core library.  I know that lots of ORM systems do indeed fetch single rows more often than they fetch batches of rows, so I can certainly imagine that it might make sense to implement it this way.

That said, thank you for taking the time to listen to my request.  I would support this if it was possible to implement.

**Comment by thomcc** [comment]
It's certainly possible to implement (As you see in the second link I gave, we implement it for `and_then` with named params (which are more than just convenience, IMO), since for us this was what we wanted 99% of the time.
